File: plugin/hooks/invoke-handler.py
#!/usr/bin/env python3
"""
invoke-handler.py - Direct handler invocation for testing and CLI use.

Usage:
    echo '{"handler": "cleanup", "context": {...}}' | python3 invoke-handler.py
    python3 invoke-handler.py '{"handler": "cleanup", "context": {...}}'

This script properly sets up the Python path to allow importing handlers
that use relative imports.
"""

import json
import sys
from pathlib import Path

# Add this directory to path BEFORE importing handlers
SCRIPT_DIR = Path(__file__).parent
sys.path.insert(0, str(SCRIPT_DIR))

from skill_handlers import get_handler


def main():
    # Read input from argument or stdin
    if len(sys.argv) > 1:
        data = json.loads(sys.argv[1])
    elif not sys.stdin.isatty():
        data = json.loads(sys.stdin.read())
    else:
        print("Usage: invoke-handler.py '<json>' or pipe JSON to stdin", file=sys.stderr)
        sys.exit(1)

    handler_name = data.get("handler")
    context = data.get("context", {})

    if not handler_name:
        print("Error: 'handler' key required in input", file=sys.stderr)
        sys.exit(1)

    handler = get_handler(handler_name)
    if not handler:
        print(f"Error: No handler registered for '{handler_name}'", file=sys.stderr)
        sys.exit(1)

    result = handler.handle(context)
    if result:
        print(result)


if __name__ == "__main__":
    main()

File: plugin/hooks/src/io/github/cowwoc/cat/hooks/InvokeHandler.java
package io.github.cowwoc.cat.hooks;

import tools.jackson.databind.JsonNode;
import tools.jackson.databind.json.JsonMapper;

/**
 * invoke-handler - Direct handler invocation for testing and CLI use.
 *
 * Usage:
 *   echo '{"handler": "cleanup", "context": {...}}' | java io.github.cowwoc.cat.hooks.InvokeHandler
 *   java io.github.cowwoc.cat.hooks.InvokeHandler '{"handler": "cleanup", "context": {...}}'
 */
public final class InvokeHandler
{
  /**
   * Entry point for direct handler invocation.
   *
   * @param args command line arguments, optionally containing JSON input
   */
  public static void main(String[] args)
  {
    try
    {
      JsonMapper mapper = JsonMapper.builder().build();
      JsonNode data;

      // Read input from argument or stdin
      if (args.length > 0)
      {
        data = mapper.readTree(args[0]);
      }
      else if (System.console() == null || System.in.available() > 0)
      {
        HookInput input = HookInput.readFromStdin();
        data = input.getRaw();
      }
      else
      {
        System.err.println("Usage: InvokeHandler '<json>' or pipe JSON to stdin");
        System.exit(1);
        return;
      }

      String handlerName = null;
      if (data.has("handler"))
      {
        handlerName = data.get("handler").asText();
      }
      JsonNode context;
      if (data.has("context"))
      {
        context = data.get("context");
      }
      else
      {
        context = mapper.createObjectNode();
      }

      if (handlerName == null || handlerName.isEmpty())
      {
        System.err.println("Error: 'handler' key required in input");
        System.exit(1);
        return;
      }

      // TODO: Port handler registry from Python
      // Handler handler = HandlerRegistry.get(handlerName);
      // if (handler == null) {
      //     System.err.println("Error: No handler registered for '" + handlerName + "'");
      //     System.exit(1);
      //     return;
      // }
      // String result = handler.handle(context);
      // if (result != null) {
      //     System.out.println(result);
      // }

      System.err.println("Error: No handler registered for '" + handlerName + "'");
      System.exit(1);
    }
    catch (Exception e)
    {
      System.err.println("Error: " + e.getMessage());
      System.exit(1);
    }
  }
}
